src/mlpack/tests/ann_layer_test.cpp

```diff
@@ -2931,4 +2931,88 @@ BOOST_AUTO_TEST_CASE(ConvolutionLayerPaddingTest)
   module2.Backward(std::move(input), std::move(output), std::move(delta));
 }
 
+/**
+ * Test that the padding options are working correctly in MaxPooling layer.
+ */
+BOOST_AUTO_TEST_CASE(MaxPoolingLayerPaddingTest)
+{
+  arma::mat output, input, delta;
+
+  // Check valid padding option.
+  MaxPooling<> module1(4, 4, 2, 2, true, std::tuple<size_t, size_t>(1, 1),
+      std::tuple<size_t, size_t>(1, 1), "valid");
+
+  // Test the Forward function.
+  input = arma::linspace<arma::colvec>(0, 24, 25);
+  module1.InputWidth() = 5;
+  module1.InputHeight() = 5;
+  module1.Forward(std::move(input), std::move(output));
+
+  BOOST_REQUIRE_EQUAL(arma::accu(output), 18);
+  BOOST_REQUIRE_EQUAL(output.n_rows, 1);
+  BOOST_REQUIRE_EQUAL(output.n_cols, 1);
+
+  // Test the Backward function.
+  module1.Backward(std::move(input), std::move(output), std::move(delta));
+
+  // Check same padding option.
+  MaxPooling<> module2(4, 4, 2, 2, true, std::tuple<size_t, size_t>(0, 0),
+      std::tuple<size_t, size_t>(0, 0), "same");
+
+  // Test the forward function.
+  input = arma::linspace<arma::colvec>(0, 24, 25);
+  module2.InputWidth() = 5;
+  module2.InputHeight() = 5;
+  module2.Forward(std::move(input), std::move(output));
+
+  BOOST_REQUIRE_EQUAL(arma::accu(output), 72);
+  BOOST_REQUIRE_EQUAL(output.n_rows, 4);
+  BOOST_REQUIRE_EQUAL(output.n_cols, 1);
+
+  // Test the backward function.
+  module2.Backward(std::move(input), std::move(output), std::move(delta));
+}
+
+/**
+ * Test that the padding options are working correctly in MeanPooling layer.
+ */
+BOOST_AUTO_TEST_CASE(MeanPoolingLayerPaddingTest)
+{
+  arma::mat output, input, delta;
+
+  // Check valid padding option.
+  MeanPooling<> module1(4, 4, 2, 2, true, std::tuple<size_t, size_t>(1, 1),
+      std::tuple<size_t, size_t>(1, 1), "valid");
+
+  // Test the Forward function.
+  input = arma::linspace<arma::colvec>(0, 24, 25);
+  module1.InputWidth() = 5;
+  module1.InputHeight() = 5;
+  module1.Forward(std::move(input), std::move(output));
+
+  BOOST_REQUIRE_EQUAL(arma::accu(output), 9);
+  BOOST_REQUIRE_EQUAL(output.n_rows, 1);
+  BOOST_REQUIRE_EQUAL(output.n_cols, 1);
+
+  // Test the Backward function.
+  module1.Backward(std::move(input), std::move(output), std::move(delta));
+
+  // Check same padding option.
+  MeanPooling<> module2(4, 4, 2, 2, true, std::tuple<size_t, size_t>(0, 0),
+      std::tuple<size_t, size_t>(0, 0), "same");
+
+  // Test the forward function.
+  input = arma::linspace<arma::colvec>(0, 24, 25);
+  module2.InputWidth() = 5;
+  module2.InputHeight() = 5;
+  module2.Forward(std::move(input), std::move(output));
+
+  BOOST_REQUIRE_EQUAL(arma::accu(output), 34.125);
+  BOOST_REQUIRE_EQUAL(output.n_rows, 4);
```

Comment:
In this test case should output be 5 x 5, same shape as input, because same padding is used here.

> With "SAME" padding, the layer's outputs will have the same spatial dimensions as its inputs.

Reply:
See [this](https://stackoverflow.com/questions/37674306/what-is-the-difference-between-same-and-valid-padding-in-tf-nn-max-pool-of-t#comment89708237_42195267). The padding convention I have used is same for the [convolution](https://stackoverflow.com/questions/37674306/what-is-the-difference-between-same-and-valid-padding-in-tf-nn-max-pool-of-t#comment89708237_42195267) one too.

Reply:
In mlpack we don't use definition of same and valid from TensorFlow as the output is same as input only when stride = 1. Here when we use "same" the output shape should be same as input input shape. 
If you look at [this,] (https://github.com/mlpack/mlpack/blob/0c09c1cb9467c846b13e641a1bdf8cab8031af40/src/mlpack/tests/ann_layer_test.cpp#L2927), Output shape is always equal to input shape when "Same" option is used.

Reply:
So this implies that if I give a 5 x 5 input and apply same padding then my output should be of size 5 x 5.

> [It looks like the output would be similar to input only if stride is equal to one.
This what actually TensorFlow does. However, we do not implement it in the similar way.
We implement such that the input is always similar to the output irrespective of stride. So it would be great if you could implement it in the same way.](https://github.com/mlpack/mlpack/issues/2119#issuecomment-568205703)